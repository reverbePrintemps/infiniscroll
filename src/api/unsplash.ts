import { createApi } from "unsplash-js";

const apiConfig = {
  accessKey: "8279f0e21833ac68af7d69e1d7c626cf0b463d8098c54e4a382b00d451ddbd22",
};

export const api = createApi(apiConfig);
