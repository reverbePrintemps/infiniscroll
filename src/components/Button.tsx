import { CSSProperties, MouseEvent, ReactNode, useState } from "react";
import "../styles/Button.css";

export interface ButtonProps {
  type?: "submit" | "reset" | "button";
  onClick?: (e: MouseEvent) => void;
  style?: CSSProperties;
  children?: ReactNode;
  className?: string;
  disabled?: boolean;
  isActive?: boolean;
  topLevel?: boolean;
  isMenu?: boolean;
  isOpen?: boolean;
}

export const Button = ({
  className,
  children,
  disabled,
  isActive,
  topLevel,
  onClick,
  isMenu,
  isOpen,
  style,
  type,
}: ButtonProps) => {
  const [isPressed, setIsPressed] = useState(false);

  return (
    <button
      className={`Button ${className || ""} ${isPressed ? "m-pressed" : ""} ${
        isActive ? "m-active" : ""
      }`}
      onClick={(e) => {
        onClick && onClick(e);
        !isMenu && e.stopPropagation();
      }}
      style={style}
      onTouchStart={() => setIsPressed(true)}
      // onTouchEnd may lead to inconsistent behavior but is necessary when scrolling over buttons on mobile
      onTouchEnd={() => setIsPressed(false)}
      onMouseDown={() => setIsPressed(true)}
      onMouseUp={() => setIsPressed(false)}
      onMouseLeave={() => setIsPressed(false)}
      disabled={disabled}
      type={type}
    >
      {!topLevel && isMenu && (
        <span className={`Button__openIndicator ${isOpen ? "m-open" : ""}`}>
          {"ğŸ‘ˆ \u00A0"}
        </span>
      )}
      {topLevel && isOpen ? <span>ğŸ™…</span> : children}
    </button>
  );
};
