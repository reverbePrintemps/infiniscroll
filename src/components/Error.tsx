import { ReactNode, CSSProperties } from "react";

interface ErrorProps {
  children?: ReactNode;
  className?: string;
  style?: CSSProperties;
}

/* No styles for Error component in this exercise. Demo purposes only. :) */
export const Error = ({ children, className, style }: ErrorProps) => {
  return (
    <div className={`Error ${className || ""}`} style={style}>
      {children}
    </div>
  );
};
