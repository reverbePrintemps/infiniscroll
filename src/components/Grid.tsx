import { ReactNode, CSSProperties } from "react";
import "../styles/Grid.css";

interface GridProps {
  children?: ReactNode;
  className?: string;
  style?: CSSProperties;
}

export const Grid = ({ children, className, style }: GridProps) => {
  return (
    <div className={`Grid ${className || ""}`} style={style}>
      {children}
    </div>
  );
};
