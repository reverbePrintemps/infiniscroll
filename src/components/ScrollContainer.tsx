import { CSSProperties, ReactNode, useEffect, useRef, useState } from "react";

import "../styles/ScrollContainer.css";

type ScrollContainerProps = {
  children: ReactNode;
  style?: CSSProperties;
  onIsReachingBottom: (bool: boolean) => void;
};

export const ScrollContainer = ({
  style,
  children,
  onIsReachingBottom,
}: ScrollContainerProps) => {
  const scrollContainerRef = useRef<HTMLDivElement>(null);
  const [windowHeight, setWindowHeight] = useState(window.innerHeight);
  const [bottomOfScrollPosition, setBottomOfScrollPosition] = useState(
    windowHeight + window.scrollY
  );
  const [bottomOfContent, setBottomOfContent] = useState<number>();
  const [isReachingBottom, setIsReachingBottom] = useState(false);

  useEffect(() => {
    if (bottomOfContent) {
      setIsReachingBottom(
        bottomOfScrollPosition >= bottomOfContent - windowHeight / 2
      );
    }
  }, [bottomOfContent, bottomOfScrollPosition, windowHeight]);

  useEffect(() => {
    if (scrollContainerRef.current && children) {
      setBottomOfContent(scrollContainerRef.current.offsetHeight);
    }
  }, [children]);

  useEffect(() => {
    onIsReachingBottom(isReachingBottom);
  }, [isReachingBottom, onIsReachingBottom]);

  const handleScroll = () => {
    if (scrollContainerRef.current) {
      setBottomOfScrollPosition(windowHeight + window.scrollY);
    }
  };

  const handleResize = () => {
    setWindowHeight(window.innerHeight);
  };

  useEffect(() => {
    window.addEventListener("scroll", handleScroll);
    window.addEventListener("resize", handleResize);
    return () => {
      window.removeEventListener("scroll", handleScroll);
      window.removeEventListener("resize", handleResize);
    };
  });

  return (
    <div ref={scrollContainerRef} className="ScrollContainer" style={style}>
      {children}
    </div>
  );
};
