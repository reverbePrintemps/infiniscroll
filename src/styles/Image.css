.Image {
  position: relative;
  min-height: 224px;
  border-radius: var(--br-card);
  box-shadow: var(--shadow-shallow);
  box-sizing: border-box;
  padding: 16px 8px;
}

.Image__overlay {
  background-color: rgba(0, 0, 0, 0.4);
  z-index: 1;
  flex-direction: column;
  /* box-sizing: border-box; */
  text-align: center;
  color: var(--c-text-primary);
  align-items: center;
  opacity: 0;
  transition: all 0.2s ease-in-out;
  min-height: 100%;
  padding: 16px 8px;
  margin: -16px -8px;
  position: relative;
  border-radius: var(--br-card);
  display: flex;
}

.Image__overlay.m-show {
  opacity: 1;
}

.Image__descriptionContainer {
  margin: auto;
}

.Image__description {
  position: relative;
  margin: 0;
  font-weight: var(--fw-extra-bold);
  word-break: break-word;
  font-size: var(--fs-s);
}

.Image__username {
  font-weight: var(--fw-regular);
  font-size: var(--fs-s);
}

.Image__favoriteButton {
  background: none;
  border: none;
  color: var(--c-text-primary);
  font-size: var(--fs-xs);
  outline: 1px solid var(--c-text-primary);
  padding: 12px 16px;
  border-radius: var(--br-pill);
  font-family: var(--ff-main);
  font-weight: var(--fw-semi-bold);
  margin: 16px 0 0;
}

.Image__favoriteButton.m-active {
  background-color: var(--c-text-primary);
  color: var(--c-text-secondary);
}

.Image__descriptionDivider {
  position: relative;
}

.Image__descriptionDivider:after {
  content: "";
  position: absolute;
  left: 50%;
  margin: 0 0 0 -50px;
  bottom: -12px;
  width: 100px;
  border-bottom: 2px solid var(--c-text-primary);
}

.Image__img {
  object-fit: cover;
  width: 100%;
  height: 100%;
  border-radius: var(--br-card);
  position: absolute;
  top: 0;
  left: 0;
}
