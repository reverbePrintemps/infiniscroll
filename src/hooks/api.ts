import { Photos } from "unsplash-js/dist/methods/search/types/response";
import { ApiResponse } from "unsplash-js/dist/helpers/response";
import { useCallback, useState } from "react";
import { api } from "../api/unsplash";

export type Photo = {
  id: number | string;
  width: number;
  height: number;
  urls: { large?: string; regular: string; raw: string; small: string };
  color: string | null;
  user: {
    username: string;
    name: string;
  };
  description: string | null;
};

const resultsPerPage = 10;

export const usePhotos = () => {
  const [photos, setPhotos] = useState<Photo[]>([]);
  const [loading, setLoading] = useState(false);
  const [error, setError] = useState<string | undefined>();
  const [searchQuery, setSearchQuery] = useState("");
  const [page, setPage] = useState(1);

  const fetchPhotos = useCallback(async (query: string, page: number) => {
    setLoading(true);
    setError(undefined);

    try {
      const res: ApiResponse<Photos> = await api.search.getPhotos({
        query,
        perPage: resultsPerPage,
        page,
      });

      if (res.response) {
        setPhotos((prevPhotos) =>
          page === 1
            ? res.response.results
            : [...prevPhotos, ...res.response.results]
        );
        setPage((prevPage) => prevPage + 1);
        setLoading(false);
      } else {
        throw new Error("No response from API");
      }
    } catch (error) {
      if (error instanceof Error) {
        setError(error.message);
      }
      setLoading(false);
    }
  }, []);

  const fetchMore = useCallback(() => {
    !loading && fetchPhotos(searchQuery, page);
  }, [loading, fetchPhotos, searchQuery, page]);

  return {
    photos,
    loading,
    error,
    fetchMore,
    fetchPhotos: (query: string) => {
      setSearchQuery(query);
      fetchPhotos(query, 1);
    },
  };
};
