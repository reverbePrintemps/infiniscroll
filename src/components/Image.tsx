import { CSSProperties, useEffect, useRef, useState } from "react";
import { HeartIcon } from "../assets/HeartIcon";
import { Skeleton } from "./Skeleton";
import { Photo } from "../hooks/api";

import "../styles/Image.css";

type ImageProps = {
  image: Photo;
  favorite: boolean;
  onClick: (id: Photo["id"]) => void;
  style?: CSSProperties;
};

export const Image = ({ image, onClick, favorite, style }: ImageProps) => {
  const [showDescription, setShowDescription] = useState(false);
  const [imageIsLoaded, setImageIsLoaded] = useState(false);
  const overlayRef = useRef<HTMLDivElement>(null);

  useEffect(() => {
    if (overlayRef.current && imageIsLoaded) {
      showDescription
        ? (overlayRef.current.style.maxHeight = `${overlayRef.current.scrollHeight}px`)
        : (overlayRef.current.style.maxHeight = "0px");
    }
  }, [showDescription, imageIsLoaded]);

  return (
    <div
      className="Image"
      onClick={() => setShowDescription(!showDescription)}
      onMouseEnter={() => setShowDescription(true)}
      onMouseLeave={() => setShowDescription(false)}
      style={style}
    >
      <div
        ref={overlayRef}
        className={`Image__overlay ${showDescription ? "m-show" : ""}`}
      >
        <div className="Image__descriptionContainer">
          {image.description && (
            <>
              <h1 className="Image__description">{image.description}</h1>
              <div className="Image__descriptionDivider"></div>
            </>
          )}
          <h2 className="Image__username">
            <i>{image.user.name}</i>
          </h2>
        </div>
        <button
          disabled={!showDescription}
          className={`Image__favoriteButton ${favorite ? "m-active" : ""}`}
          onClick={(e) => {
            onClick && onClick(image.id);
            e.stopPropagation();
          }}
        >
          Favorite
        </button>
      </div>
      {favorite && <HeartIcon className="Image__favoriteIcon" />}
      {!imageIsLoaded && <Skeleton className="Image__img" />}
      <img
        className={`Image__img ${imageIsLoaded ? "m-show" : ""}}`}
        src={image.urls.small}
        alt={image.description || ""}
        onLoad={() => setImageIsLoaded(true)}
        // 93.05% global coverage as of this writing
        loading="lazy"
      />
    </div>
  );
};
