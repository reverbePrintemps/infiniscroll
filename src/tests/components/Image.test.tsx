import { render, fireEvent, screen, waitFor } from "@testing-library/react";
import { Image } from "../../components/Image";
import { Photo } from "../../hooks/api";
import { act } from "react-dom/test-utils";
import userEvent from "@testing-library/user-event";

const mockImage: Photo = {
  id: 1,
  urls: {
    small: "small_image_url",
    regular: "regular_image_url",
    raw: "raw_image_url",
  },
  description: "Test image",
  user: {
    name: "John Doe",
    username: "johndoe",
  },
  width: 100,
  height: 100,
  color: "blue",
};

describe("Image component", () => {
  test("renders without errors", () => {
    render(
      <Image
        image={mockImage}
        onClick={jest.fn()}
        favorite={false}
        style={{}}
      />
    );

    const imageElement = screen.getByAltText(mockImage.description as string);
    expect(imageElement).toBeInTheDocument();
  });

  test("toggles description on click", () => {
    render(
      <Image
        image={mockImage}
        onClick={jest.fn()}
        favorite={false}
        style={{}}
      />
    );

    const descriptionContainer = screen.getByTestId(
      "image-description-container"
    );
    fireEvent.click(descriptionContainer);

    expect(descriptionContainer).toMatchSnapshot();

    fireEvent.click(descriptionContainer);

    expect(descriptionContainer).toMatchSnapshot();
  });

  test("displays favorite icon when favorite prop is true", () => {
    render(
      <Image image={mockImage} onClick={jest.fn()} favorite={true} style={{}} />
    );

    const favoriteIcon = screen.getByTestId("FavoriteIcon");
    expect(favoriteIcon).toBeInTheDocument();
  });

  test("displays skeleton when image is not loaded", () => {
    render(
      <Image
        image={mockImage}
        onClick={jest.fn()}
        favorite={false}
        style={{}}
      />
    );

    const skeletonElement = screen.getByTestId("image-skeleton");
    expect(skeletonElement).toBeInTheDocument();
  });

  test("displays loaded image when image is loaded", () => {
    render(
      <Image
        image={mockImage}
        onClick={jest.fn()}
        favorite={false}
        style={{}}
      />
    );

    const imageElement = screen.getByAltText(mockImage.description as string);
    fireEvent.load(imageElement);

    expect(imageElement).toBeInTheDocument();
  });
});
