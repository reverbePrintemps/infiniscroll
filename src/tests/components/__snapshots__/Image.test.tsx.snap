// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Image component toggles description on click 1`] = `
<div
  class="Image"
  data-testid="image-description-container"
>
  <div
    class="Image__overlay m-show"
  >
    <div
      class="Image__descriptionContainer"
    >
      <h1
        class="Image__description"
      >
        Test image
      </h1>
      <div
        class="Image__descriptionDivider"
      />
      <h2
        class="Image__username"
      >
        <i>
          John Doe
        </i>
      </h2>
    </div>
    <button
      class="Image__favoriteButton "
    >
      Favorite
    </button>
  </div>
  <div
    class="Skeleton Image__img"
    data-testid="image-skeleton"
  />
  <img
    alt="Test image"
    class="Image__img }"
    loading="lazy"
    src="small_image_url"
  />
</div>
`;

exports[`Image component toggles description on click 2`] = `
<div
  class="Image"
  data-testid="image-description-container"
>
  <div
    class="Image__overlay "
  >
    <div
      class="Image__descriptionContainer"
    >
      <h1
        class="Image__description"
      >
        Test image
      </h1>
      <div
        class="Image__descriptionDivider"
      />
      <h2
        class="Image__username"
      >
        <i>
          John Doe
        </i>
      </h2>
    </div>
    <button
      class="Image__favoriteButton "
      disabled=""
    >
      Favorite
    </button>
  </div>
  <div
    class="Skeleton Image__img"
    data-testid="image-skeleton"
  />
  <img
    alt="Test image"
    class="Image__img }"
    loading="lazy"
    src="small_image_url"
  />
</div>
`;
